lib/logger/lib/logger/config.ex

```diff
@@ -139,69 +143,147 @@ defmodule Logger.Config do
     {:ok, state}
   end
 
-  ## Helpers
+  defp update_counter({counter, log, discard_threshold, discard_period}, periodic_check?) do
+    # If length is more than the total, it means the counter is behind,
+    # due to non-log messages, so we need to increase the counter.
+    #
+    # If length is less than the total, then we either have a spike or
+    # the counter drifted due to failures.
+    #
+    # Because we always bump the counter and then we send the message,
+    # there is a chance clients have bumped the counter but they did not
+    # deliver the message yet. Those bumps will be lost. At the same time,
+    # we are careful to read the counter first here, so if the counter is
+    # bumped after we read from it, those bumps won't be lost.
+    total = read_counter(counter)
+    {:message_queue_len, length} = Process.info(self(), :message_queue_len)
+    add_counter(counter, length - total)
+
+    # In case we are logging but we reached the threshold, we log that we
+    # started discarding messages. This can only be reverted by the periodic
+    # discard check.
+    cond do
+      total >= discard_threshold ->
+        if log == :log or periodic_check? do
+          warn("Attempted to log #{total} messages, which is above :discard_threshold")
+        end
+
+        {counter, :discard, discard_threshold, discard_period}
+
+      log == :discard and periodic_check? ->
+        warn("Attempted to log #{total} messages, which is below :discard_threshold")
```

Comment:
There isnt a test for this feature

Reply:
I tried but it is really tricky to test it without adding more complexity and/or adding long timers to the suite.

Reply:
I think I can test it by manually sending  a `handle_info`. Not nice but it should do it. 🤷‍♂️

Reply:
Sounds good